### ISSUE TITLE
Building DALES dataset

### ISSUE THREAD

**Issue by eyalol** [HuguesTHOMAS/KPConv-PyTorch#72]
Hello Thomas,
First of all, thanks for this amazing open-sourced code. 
I'm embarking on a project whose goal is to train KPConv on the DALES dataset,  As part of this effort, first, a dales dataset class is needed, which I want to make similar to your S3DIS dataset class. 
Currently, I'm digging into your code, and I'd like to know from you which parts of this dataset Class require change for the dales dataset. For example, is the sampler needs change? etc. 
Thanks!

**Comment by HuguesTHOMAS** [comment]
Hi @eyalol,

Thanks for your interest in my code. If you want, you can have a look at [this repository](https://github.com/Arjun-NA/KPConv_for_DALES), where KPConv is applied on DALES dataset.

If you want to restart the implementation from scratch, then my advice is to keep as much as you can from the S3DIS dataset class. Just replace the name S3DIS by DALES, and the path to the data. Then you just have to create a small script that converts the DALES Ascii file to binary ply files and store them in the same folder structure as S3DIS.

Good luck with your project.
HUgues

**Comment by eyalol** [comment]
@HuguesTHOMAS  Thanks for the fast response, I've looked into this repo, and saw that it was implemented with TensorFlow. I want to use the PyTorch version.

So you're saying I could basically just use the S3DIS file for the DALES dataset as well, of course after converting the DALES dataset to the same folder structure as S3DIS. Though it seems odd to me that no further changes are needed? I'm sure I'll notice changes (if needed) when I run the code, but i'd like to re-assure that every piece of code should work for the dales dataset too, for example, DALES doesn't have color features, in contrast to the S3DIS, so I'm assuming it could cause significant changes to the code. 
Thank you.

**Comment by HuguesTHOMAS** [comment]
@eyalol,

Yes indeed. You will have to modify the part of the code which loads the point clouds (in the __get_item__) function of the dataset. This function has two possible behavior: potential item and random_item but the part where you load the pointclouds are identical:

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/bb0ea445b168463d4b7dffee8b61c235775fc33a/datasets/S3DIS.py#L343-L350

and 

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/bb0ea445b168463d4b7dffee8b61c235775fc33a/datasets/S3DIS.py#L536-L543

**Comment by HuguesTHOMAS** [comment]
From these two points, you can follow the code and modify every time it is needed until the end of the __get_item__ function. The rest of the dataset should not need a modification I believe.

Best,
Hugues

**Comment by eyalol** [comment]
@HuguesTHOMAS Hello again Hugues, 
I've started working on the modification for the S3DIS to work on dales, as you've mentioned, I've noticed two possible behaviors, random item and potential item, the potential item method confuses me a bit. could you provide some information about it? a source to read about it would be excellent too. 

P.S I'm assuming that in the process of developing the dataset I'll have more questions, is there any other way of communicating with you? discord etc.. 

Thanks!

**Comment by eyalol** [comment]
It will be lovely if you will provide some information about what each item's role is. 
i'd like after finishing to document everything and push to this repo ofcourse. 

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/bb0ea445b168463d4b7dffee8b61c235775fc33a/datasets/S3DIS.py#L495-L505

**Comment by HuguesTHOMAS** [comment]
Hi @eyalol,

First, you can read about potential sampling strategy in my thesis manuscript [here](https://pastel.archives-ouvertes.fr/tel-02458455/document), section IV.2.8.c (p127-128). I think I also mention it and gave a short explanation in another issue of this repository.

You can post your questions here, I will generally answer pretty fast, and I like to have all the conversation about the method logged here for people to find if needed.

Concerning the lists, they are containers for each batch element. 

     p_list = []   => input points (N, 3) float array
     f_list = []    => input features (N, D) float array
     l_list = []    => input labels (N,) int array
     pi_list = []  => indices of the points inside the input sphere (N,) int array
     i_list = []    => index of the point chosen as center of the input sphere (integer)
     ci_list = []  => index of the pointcloud where the input is picked (integer)
     s_list = []   => augmentation scale applied to the input points (float or 3 floats if anysotropic scale)
     R_list = []  => augmentation rotation applied to the input points (3 by 3 matrix)

You can see that just below in the code:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/bb0ea445b168463d4b7dffee8b61c235775fc33a/datasets/S3DIS.py#L555-L563

They are not all necessary for the network, some of these are just here for debugging purpose of for the validation/test of the model.

Best,
Hugues

**Comment by eyalol** [comment]
I see. Did you notice any significant performance change between random item and potential item? 
I'd also like to ask what features are supposed to be accepted by the network, I see here: https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/bb0ea445b168463d4b7dffee8b61c235775fc33a/datasets/S3DIS.py#L591-L600

That either 1, 4, 7 dimensions features are accepted, what does it cope with? 

Thank you.

**Comment by HuguesTHOMAS** [comment]
> Did you notice any significant performance change between random item and potential item

For training, there is not much difference, but for validation and test, it is better to use the potential item to ensure you cover all the area you want to predict and don't pick more items where they have already been picked

> I'd also like to ask what features are supposed to be accepted by the network

You can modify these lines of the code if you want. I usually adapt these depending on what the dataset has to offer. You can add any value you want to the network features, _height-above-ground_ for example can be relevant for an airborne lidar dataset, _color_ if you have it, or the _intensity_ of the lidar. My advice is to always keep the first column of ones (this is important) and optionally add other features that you find relevant. Just make sure that the `self.config.in_features_dim` is actually the dimension of the features you use, otherwise, it will crash.

**Comment by eyalol** [comment]
> > Did you notice any significant performance change between random item and potential item
> 
> For training, there is not much difference, but for validation and test, it is better to use the potential item to ensure you cover all the area you want to predict and don't pick more items where they have already been picked
> 
> > I'd also like to ask what features are supposed to be accepted by the network
> 
> You can modify these lines of the code if you want. I usually adapt these depending on what the dataset has to offer. You can add any value you want to the network features, _height-above-ground_ for example can be relevant for an airborne lidar dataset, _color_ if you have it, or the _intensity_ of the lidar. My advice is to always keep the first column of ones (this is important) and optionally add other features that you find relevant. Just make sure that the `self.config.in_features_dim` is actually the dimension of the features you use, otherwise, it will crash.

What is the 1's feature column role?

**Comment by HuguesTHOMAS** [comment]
It is explained in the paper. a feature equal to zero is mathematically equivalent to empty space. So to avoid that point zero features (like a black colored area) are ignored by the network, you have to give a feature equal to one of all the points. You can see it as the binary value in a grid 0 is empty and one is full. This column of one thus encoded the geometry of the shapes

**Comment by Dhagash4** [comment]
@HuguesTHOMAS First of all thank you for making your code open-source, I am also trying to run your code on my dataset, I just changed the ply file creation in S3DIS dataset and its working, I just have points and labels in my dataset but to make it consistent with S3DIS dataset I have added dummy colors but that information is not valuable during training so I was confused as to what input features I have to select I am currently using 4,Is that fine? Also can you share some possible directions which parameters to change during training, I saw your comments and I am changing in_radius and first_subsampling_dl but sometimes it gives me error like "Expected  more than one value per channel when training" what can be the possible cause?

Thank You for your help your comments really helped.

**Comment by HuguesTHOMAS** [comment]
Hi @Dhagash4,

You can set the parameter `in_features_dim` to 1. It is used there:

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/1cd3742fc57e5ba9d21f7909c78a05a971781351/datasets/S3DIS.py#L402-L411

As you can see, a value of 1 will have the code ignore colors. Alternatively, you can modify this piece of code if you want to add specific features to your data, like height above ground for example. Just create a "elif" case like this 

```
elif self.config.in_features_dim == (1 + your_size):
    stacked_features = np.hstack((stacked_features, your_features))
```

Where `your_features` is an array of size `(N, your_size)`.

To answer you other question, you can play with a lot of thing:

 - `architecture`: play with the number of resnet layer, etc.
 - `in_radius` and `first_subsampling_dl`: play with the scale and density of the input
 - `first_features_dim`: number of feature channels in the network. This is the value at the first layer and at each layer the number of features is multiplied by two.
 - `learning_rate` and `lr_decays`: control the convergence of the network, smaller decay will make the convergence slower but will also make sure that you reach the minimum loss, sometimes decay too high means the network converges before reaching the minimum. 
 - `augment_XXX`: Controls the data augmentation process, random rotations, scaling, symmetries, noise, etc.

You can also play with other parameters if you want but these ones are more sensible:
 - `conv_radius`: play with the scale of the convolution kernel with respect to data density (larger means more computations)
 - `KP_extent`, `KP_influence`, `aggregation_mode`: these ones control the nature of our convolution kernel


Eventually, the error you mention seems to come from the batch normalization when the point clouds only contain 1 point. I never had this one, but the solution seems pretty simple to me, I assume it happened when you used a smaller `in_radius`, and you were only left with one point in the latest layer, so you could either keep larger `in_radius`, or reduce the number of layers in the network. Furthermore, if you have isolated points in your dataset, you should probably remove them to avoid such errors.

**Comment by amirhszd** [comment]
Hi @eyalol, I want to use KPconv with DALES as well as the project I am working on requires Aerial dataset. Do you think it would be possible for us to get in touch?

Thanks, Amir
